crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1392,48 +1477,137 @@ impl<'db> Type<'db> {
+            Type::StringLiteral(_) | Type::LiteralString => {
+                KnownClass::Str.to_instance(db).static_member(db, name)
             }
 
-            Type::BytesLiteral(_) => KnownClass::Bytes.to_instance(db).member(db, name),
+            Type::BytesLiteral(_) => KnownClass::Bytes.to_instance(db).static_member(db, name),
 
             // We could plausibly special-case `start`, `step`, and `stop` here,
             // but it doesn't seem worth the complexity given the very narrow range of places
             // where we infer `SliceLiteral` types.
-            Type::SliceLiteral(_) => KnownClass::Slice.to_instance(db).member(db, name),
+            Type::SliceLiteral(_) => KnownClass::Slice.to_instance(db).static_member(db, name),
 
-            Type::Tuple(_) => {
-                // TODO: implement tuple methods
-                Symbol::todo("Attribute access on heterogeneous tuple types")
-            }
+            Type::Tuple(_) => KnownClass::Tuple.to_instance(db).static_member(db, name),
 
             Type::AlwaysTruthy | Type::AlwaysFalsy => match name {
                 "__bool__" => {
                     // TODO should be `Callable[[], Literal[True/False]]`
                     Symbol::todo("`__bool__` for `AlwaysTruthy`/`AlwaysFalsy` Type variants")
                 }
-                _ => Type::object(db).member(db, name),
+                _ => Type::object(db).static_member(db, name),
             },
         }
     }
 
+    /// Call the `__get__(instance, owner)` method on a type, if it exists.
+    fn try_call_dunder_get(
+        &self,
+        db: &'db dyn Db,
+        instance: Option<Type<'db>>,
+        owner: Type<'db>,
+    ) -> Option<Type<'db>> {
+        // TODO: Handle possible-unboundness and errors from `__get__` calls.
+        self.member(db, "__get__")
+            .ignore_possibly_unbound()
+            // TODO:
+            // We currently filter out `Unknown` here because the fallout would be too large otherwise.
+            // Imagine that we access a class attribute `C.attr`, and the static member lookup of `attr`
+            // returns `Unknown | T` (something we do for all undeclared attributes). Now we do a static
+            // lookup of the `__get__` method on that `Unknown | T` type. For `Unknown`, we get a bound
+            // symbol with type `Unknown` back. If `T` is not a descriptor, we get back that `__get__`
+            // is *unbound* on `T`. We then union `Symbol::bound(Unknown)` and `Symbol::unbound()` into
+            // a possibly-unbound symbol of type `Unknown`. If we would not filter it out here, that
+            // would mean that we would infer `Unknown` for all of these attribute accesses. A better
+            // way to handle this would be to return something like a `Symbol::union` of the two results,
+            // and then to do a `__get__` call on `Unknown`, and to handle `T` as a non-descriptor member,
+            // resulting in `Unknown | T`.
```

Comment:
I do think this describes what we need to do correctly, but I'm still not totally convinced that embedding this in `Symbol` is the best implementation approach, vs abstracting the concept of mapping an operation (expressed as a callback) over a type (which may be a union or intersection). This avoids having to package up a union or intersection of symbols inside `Symbol` (making it another representation of union or intersection outside Type).

Reply:
This comment is now removed, and we properly support unions by matching on them explicitly. This was implemented in a fraction of the time that it took me to write this comment :upside_down_face:.